### PR TITLE
Replace jQuery code with vanilla JS for better performance

### PR DESCRIPTION
**Pull request by rafaucau**
**Changes proposed in this pull request:**

- ~~Bundle `pusher-js`. Instead of loading the Pusher script with `getScript`, it is now included in the extension. This gives us fewer HTTP requests and also makes the code of this extension smaller, since it no longer requires promises to be used every time Pusher is called.~~
- Remove code that is no longer used.
- ~~Update a very old version of `pusher/pusher-php-server`.~~
- Replace jQuery code with vanilla JS for better performance

**Necessity**

- [x] Has the problem that is being solved here been clearly explained?
- [x] If applicable, have various options for solving this problem been considered?
- [x] For core PRs, does this need to be in core, or could it be in an extension?
- [x] Are we willing to maintain this for years / potentially forever?

**Confirmed**

- [x] Frontend changes: tested on a local Flarum installation.
- [x] Backend changes.
- [ ] Core developer confirmed locally this works as intended.
- [x] Tests have been added, or are not appropriate here.

### CODE REVIEW

**Comment by dsevillamartin** [comment]
I personally am not in favor of bundling pusher js in the extension.

I consider pusher to not be necessary or q priority to a forums functionality. I think I'd rather have extensions load it when needed than always be loaded.

**Comment by tankerkiller125** [comment]
I'm going to agree with @datitisev that bundling the Pusher JS is not ideal. It's optional functionality that does not impact overall operations of the forum if it fails to load or operate. If the CDN is failing to load in a certain country or something of that nature then we should look into maybe switching it to a provider that will work in said country.

**Comment by rafaucau** [comment]
@datitisev @tankerkiller125 Ok, I have reverted the changes. 
It's worth replacing `$.getScript` with something else anyway, since this function disables script caching.

**Comment by rafaucau** [comment]
After replacing `$.getScript` with custom JS:
image

Pity that the Flarum configuration does not support native `import()`.

**Comment by dsevillamartin** [comment]
Am I correct in stating the update to pusher/pusher-php-server probably introduces breaking changes to extensions that integrate Pusher? We're jumping 5 major versions, and in this diff alone I see a change from `get_channels` into `getChannels`.

> Pity that the Flarum configuration does not support native import().

Fairly certain it does? I used it in `fof/discussion-language` at one point (https://github.com/FriendsOfFlarum/discussion-language/blob/5e9b8a60e4ef7b64eaffcace1723537f2eab367b/js/src/admin/utils/countries.js#L8). And if not, I think the current code is sufficient (the one you wrote) but am not entirely sure.

**Comment by rafaucau** [comment]
> @datitisev Am I correct in stating the update to pusher/pusher-php-server probably introduces breaking changes to extensions that integrate Pusher? We're jumping 5 major versions, and in this diff alone I see a change from get_channels into getChannels.

Yes, the latest version of Pusher has camelCase method names instead of snake_case. But updating Pusher is worth doing anyway. It is surprising that such an old version works with PHP 8.

> Fairly certain it does? I used it in fof/discussion-language at one point (https://github.com/FriendsOfFlarum/discussion-language/blob/5e9b8a60e4ef7b64eaffcace1723537f2eab367b/js/src/admin/utils/countries.js#L8). And if not, I think the current code is sufficient (the one you wrote) but am not entirely sure.

Oh, so it is possible then. My solution works, but the one with `import()` seems more elegant.

**Comment by dsevillamartin** [comment]
> Yes, the latest version of Pusher has camelCase method names instead of snakeCase. But updating Pusher is worth doing anyway. It is surprising that such an old version works with PHP 8.

I don't disagree with this, but this introduces breaking changes for extensions, which we cannot do unless we release a major version. Not sure what our thoughts are on releasing major versions for extensions only -- that'd require more discussion.

> Oh, so it is possible then. My solution works, but the one with import() seems more elegant.

Yep, looks like this code can be simplified to something like
```js
app.pusher = import('<link>')
  .then(() => 
    // the onload stuff
  })
```

Now that I look at it, it seems pusher is not loaded when needed - it's loaded immediately because the body of `new Promise` runs instantly... huh 🤔

**Comment by davwheat** [comment]
> Not sure what our thoughts are on releasing major versions for extensions only -- that'd require more discussion.

As far as I'm aware, we're trying to keep extension major versions in sync with core, meaning breaking updates aren't really an option (unless we obviously really need to, for security reasons or similar).

We can keep a second PR open for the version bump and merge for Flarum 2.0, though.

**Comment by rafaucau** [comment]
Ok, I have reverted the backend update.

> @datitisev Now that I look at it, it seems pusher is not loaded when needed - it's loaded immediately because the body of new Promise runs instantly... huh 🤔

Pusher is always loaded:
https://github.com/flarum/pusher/blob/3b1af6dfbeaf7fc8329face0abf85e1e68bce4c8/js/src/forum/index.ts#L152
This was before my changes.

---
Anyway, the way with `import()` now works. `webpackIgnore: true` is needed for this to work.

File: js/src/forum/index.ts [hunk 1]
```diff
@@ -131,7 +112,7 @@ app.initializers.add('flarum-pusher', () => {
             if (!document.hasFocus()) {
               app.setTitleCount(Math.max(0, this.discussion.commentCount() - oldCount));
 
-              $(window).one('focus', () => app.setTitleCount(0));
+              window.addEventListener('focus', () => app.setTitleCount(0));
```
Comment (dsevillamartin) [inline js/src/forum/index.ts]: Need to add the `once` option to this to keep the behavior the same.  `one` makes the callback be called for the 1st event per element, and here it's the same as once.

File: js/tsconfig.json [hunk 2]
```diff
@@ -7,6 +7,7 @@
     "baseUrl": ".",
     "paths": {
       "flarum/*": ["../vendor/flarum/core/js/dist-typings/*"]
-    }
+    },
+    "module": "ES2020"
```
Comment (askvortsov1) [inline js/tsconfig.json]:
```suggestion
    "module": "es2020"
```

By convention I think we keep this lowercase.

@davwheat I wonder if we should bump our TSConfig to use newer versions of JS for target / module?

**Review by askvortsov1 — APPROVED** [review]

**Comment by askvortsov1** [comment]
Thank you for the PR!

File: js/src/forum/index.ts [hunk 3]
```diff
@@ -9,25 +9,25 @@ import ItemList from 'flarum/common/utils/ItemList';
...
-  app.pusher = new Promise((resolve) => {
-    $.getScript('//cdn.jsdelivr.net/npm/pusher-js@7.0.3/dist/web/pusher.min.js', () => {
-      const socket: PusherTypes.default = new Pusher(app.forum.attribute('pusherKey'), {
-        authEndpoint: `${app.forum.attribute('apiUrl')}/pusher/auth`,
-        cluster: app.forum.attribute('pusherCluster'),
-        auth: {
-          headers: {
-            'X-CSRF-Token': app.session.csrfToken,
-          },
+  app.pusher = new Promise(async (resolve) => {
```
Comment (davwheat) [inline js/src/forum/index.ts]: Using `new Promise()` and `async () => {}` is an anti-pattern. We should use either one or the other. (I'd prefer an async function.)

In this example...

```ts
const a = new Promise(async (resolve) => {
  resolve(await someAsyncCode());
});
```

can be simplified to...

```ts
const a = async () => {
  return await someAsyncCode();
}
```

**Review by davwheat — CHANGES_REQUESTED** [review]

File: js/src/forum/index.ts (hunk 3 above)
Comment (rafaucau) [reply js/src/forum/index.ts]: @davwheat I'm probably not thinking right at this late hour, but something like this doesn't work:
```ts
  app.pusher = async () => {
    await import('//cdn.jsdelivr.net/npm/pusher-js@7.0.3/dist/web/pusher.min.js' /* webpackIgnore: true, webpackPrefetch: true */);

    const socket: PusherTypes.default = new Pusher(app.forum.attribute('pusherKey'), {
      authEndpoint: `${app.forum.attribute('apiUrl')}/pusher/auth`,
      cluster: app.forum.attribute('pusherCluster'),
      auth: {
        headers: {
          'X-CSRF-Token': app.session.csrfToken,
        },
      },
    });

    return {
      channels: {
        main: socket.subscribe('public'),
        user: app.session.user ? socket.subscribe(`private-user${app.session.user.id()}`) : null,
      },
      pusher: socket,
    };
  };
```

Comment (davwheat) [reply js/src/forum/index.ts]: Ah sorry, my typo.

It can be simplified to:

```ts
const a = (async () => {
  return await someAsyncCode();
})()
```

**Review by davwheat — APPROVED** [review]

**Pull request merged**